extensions/components/com_redcore/admin/controllers/webservice.php

```diff
@@ -29,26 +29,68 @@ public function ajaxGetTask()
 		$input = $app->input;
 
 		$taskName = $input->getString('taskName', '');
+
+		if (empty($taskName))
+		{
+			$app->close();
+		}
+
 		$model = $this->getModel();
 		$model->formData['task-' . $taskName] = $model->bindPathToArray('//operations/taskResources', $model->defaultXmlFile);
 		$model->setFieldsAndResources('task-' . $taskName, '//operations/taskResources', $model->defaultXmlFile);
 
-		if (!empty($taskName))
-		{
-			echo RLayoutHelper::render(
-				'webservice.operation',
-				array(
-					'view' => $model,
-					'options' => array(
-						'operation' => 'task-' . $taskName,
-						'form'      => $model->getForm($model->formData, false),
-						'tabActive' => ' active in ',
-						'fieldList' => array('defaultValue', 'isRequiredField', 'isPrimaryField'),
-					)
+		echo RLayoutHelper::render(
+			'webservice.operation',
+			array(
+				'view' => $model,
+				'options' => array(
+					'operation' => 'task-' . $taskName,
+					'form'      => $model->getForm($model->formData, false),
+					'tabActive' => ' active in ',
+					'fieldList' => array('defaultValue', 'isRequiredField', 'isPrimaryField'),
 				)
-			);
+			)
+		);
+
+		$app->close();
+	}
+
+	public function ajaxAddComplexType()
+	{
+		$app = JFactory::getApplication();
+		$input = $app->input;
+
+		$typeName = $input->getString('typeName', '');
+
+		if (empty($typeName))
+		{
+			$app->close();
 		}
 
+		/** @var RedcoreModelWebservice $model */
+		$model = $this->getModel();
+		$xml = simplexml_load_string(str_replace('"operation"', '"type-' . $typeName . '"', $model->loadFormComplexTypeXml()));
+
+		$model->formData['type-' . $typeName] = $model->bindPathToArray('//complexArrays/' . $typeName, $xml);
+		$model->setPropertyByXpath('fields', 'type-' . $typeName, '//complexArrays/' . $typeName . '/fields/field', $xml);
+
+		$view = $this->getView('Webservice', 'html');
+		$view->setModel($model, true);
+
+
```

Comment:
I am surprised that code sniffer is not complaining about double empty end line spaces, this is not such an issue, but  it would be good that you fix those with the rest of my comments...